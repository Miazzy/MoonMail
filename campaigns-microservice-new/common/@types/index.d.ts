import {APIGatewayEventRequestContext, APIGatewayProxyEvent} from 'aws-lambda';

export type TokenPayload = {
  aud: string;
  email: string;
  email_verified: boolean;
  exp: number;
  iat: number;
  iss: string;
  sub: string;
  [key: string]: any;
};

export type AuthorizerContext = {
  userId: string;
  principalId: string;
  email?: string;
};

export interface AuthorizedEventRequestContext extends APIGatewayEventRequestContext {
  authorizer: AuthorizerContext;
}

export interface AuthorizedEvent extends APIGatewayProxyEvent {
  user?: AuthorizerContext;
  requestContext: AuthorizedEventRequestContext;
}

export interface NormalizedEvent {
  user: AuthorizerContext;
  headers: {[name: string]: string};
  httpMethod: string;
  isBase64Encoded: boolean;
  path: string;
  pathParameters: {
    [name: string]: string,
    fields: Array<string>
  } | null;
  queryStringParameters: {
    [name: string]: string|number|Array<string>,
    page: number,
    limit: number,
    fields: Array<string>,
    archived: boolean
  } | null,
  stageVariables: {[name: string]: string} | null;
  requestContext: APIGatewayEventRequestContext;
  resource: string;
  body: object;
}

export type Auth0UserData = {
  user_id: string;
  email_verified: boolean;
  email: string;
  picture: string;
  nickname: string;
  user_metadata: {[name: string]: any};
  app_metadata: {[name: string]: any};
};

export type Response = {
  statusCode?: number
  [key: string]: any
}