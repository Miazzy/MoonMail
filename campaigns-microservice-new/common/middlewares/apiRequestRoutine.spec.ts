import {
  addAuthorizerContext, apiRequestRoutine, buildErrorResponse,
  buildJSONResponse,
  setDefaultParams
} from './apiRequestRoutine';
import {NotFound} from 'http-errors';
import {AuthorizedEvent, NormalizedEvent} from '../@types';
import middy from 'middy';

describe('apiRequestRoutine', () => {
  let demoEvent: AuthorizedEvent;
  beforeEach(() => {
    demoEvent = {
      resource: '/account',
      path: '/account',
      httpMethod: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: 'valid_token'
      },
      queryStringParameters: null,
      pathParameters: null,
      stageVariables: null,
      requestContext: {
        resourceId: 'una98p',
        authorizer: {
          principalId: 'auth0|56cf10321c4c7f2e2fb264aa',
          userId: 'auth0|56cf10321c4c7f2e2fb264aa'
        },
        resourcePath: '/account',
        httpMethod: 'POST',
        path: '/dev/account',
        accountId: '862946620411',
        stage: 'dev',
        requestTimeEpoch: 1539156810614,
        requestId: 'c901f211-cc5e-11e8-b7a5-4bcd36c6e6f3',
        identity: {
          apiKey: null,
          apiKeyId: null,
          cognitoIdentityPoolId: null,
          accountId: null,
          cognitoIdentityId: null,
          caller: null,
          sourceIp: '35.167.74.121',
          accessKey: null,
          cognitoAuthenticationType: null,
          cognitoAuthenticationProvider: null,
          userArn: null,
          userAgent: 'axios/0.15.2',
          user: null
        },
        apiId: 'q27n90kff1'
      },
      body: "{\"foo\":\"bar\"}",
      isBase64Encoded: false
    };
  });

  it('addAuthorizerContext should assign user object to the event', () => {
    addAuthorizerContext(demoEvent);
    expect(demoEvent.user).toEqual(demoEvent.requestContext.authorizer)
  });

  it('setDefaultParams should set default query and path parameters', () => {
    setDefaultParams(demoEvent);
    expect(demoEvent.queryStringParameters).toEqual({});
    expect(demoEvent.pathParameters).toEqual({});
  });

  it('buildJSONResponse should build a correct response object', () => {
    expect(buildJSONResponse({foo: 'bar'})).toMatchSnapshot()
  });

  it('buildErrorResponse should build a correct response object in case of the internal error', () => {
    expect(buildErrorResponse(new Error('Some internal error'))).toMatchSnapshot()
  });

  it('buildErrorResponse should build a correct response object in case of the http error', () => {
    expect(buildErrorResponse(new NotFound('Recourse not found'))).toMatchSnapshot()
  });

  it('should normalize event', () => {
    const handler = middy(async(event: NormalizedEvent) => {
      expect(event).toMatchSnapshot();
    });
    handler.use(apiRequestRoutine());

    handler(demoEvent, <any>{}, () => {});
  });

  it('should return success response', () => {
    const handler = middy(async() => {
      return {foo: 'bar'}
    });
    handler.use(apiRequestRoutine());

    handler(demoEvent, <any>{}, (_, response) => {
      expect(response).toMatchSnapshot();
    })
  });

  it('should return error response if http error is thrown in handler', () => {
    const handler = middy(async() => {
      throw new NotFound('Recourse not found');
    });
    handler.use(apiRequestRoutine());

    handler(demoEvent, <any>{}, (_, response) => {
      expect(response).toMatchSnapshot();
    })
  });

  it('should return error response if internal error is thrown in handler', () => {
    const handler = middy(async() => {
      throw new Error('Some internal error');
    });
    handler.use(apiRequestRoutine());

    // @ts-ignore
    handler(demoEvent, {}, (_, response) => {
      expect(response).toMatchSnapshot();
    })
  });
});
