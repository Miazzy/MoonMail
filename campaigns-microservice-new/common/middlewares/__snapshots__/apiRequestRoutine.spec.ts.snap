// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiRequestRoutine buildErrorResponse should build a correct response object in case of the http error 1`] = `
Object {
  "body": "{\\"status\\":\\"NotFoundError\\",\\"statusCode\\":404,\\"message\\":\\"Recourse not found\\"}",
  "statusCode": 404,
}
`;

exports[`apiRequestRoutine buildErrorResponse should build a correct response object in case of the internal error 1`] = `
Object {
  "body": "{\\"status\\":\\"ServiceUnavailableError\\",\\"statusCode\\":503,\\"message\\":\\"Upstream service is unavailable\\"}",
  "statusCode": 503,
}
`;

exports[`apiRequestRoutine buildJSONResponse should build a correct response object 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`apiRequestRoutine should normalize event 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "valid_token",
    "Content-Type": "application/json",
  },
  "httpMethod": "POST",
  "isBase64Encoded": false,
  "path": "/account",
  "pathParameters": Object {},
  "queryStringParameters": Object {},
  "requestContext": Object {
    "accountId": "862946620411",
    "apiId": "q27n90kff1",
    "authorizer": Object {
      "principalId": "auth0|56cf10321c4c7f2e2fb264aa",
      "userId": "auth0|56cf10321c4c7f2e2fb264aa",
    },
    "httpMethod": "POST",
    "identity": Object {
      "accessKey": null,
      "accountId": null,
      "apiKey": null,
      "apiKeyId": null,
      "caller": null,
      "cognitoAuthenticationProvider": null,
      "cognitoAuthenticationType": null,
      "cognitoIdentityId": null,
      "cognitoIdentityPoolId": null,
      "sourceIp": "35.167.74.121",
      "user": null,
      "userAgent": "axios/0.15.2",
      "userArn": null,
    },
    "path": "/dev/account",
    "requestId": "c901f211-cc5e-11e8-b7a5-4bcd36c6e6f3",
    "requestTimeEpoch": 1539156810614,
    "resourceId": "una98p",
    "resourcePath": "/account",
    "stage": "dev",
  },
  "resource": "/account",
  "stageVariables": null,
  "user": Object {
    "principalId": "auth0|56cf10321c4c7f2e2fb264aa",
    "userId": "auth0|56cf10321c4c7f2e2fb264aa",
  },
}
`;

exports[`apiRequestRoutine should return error response if http error is thrown in handler 1`] = `
Object {
  "body": "{\\"status\\":\\"NotFoundError\\",\\"statusCode\\":404,\\"message\\":\\"Recourse not found\\"}",
  "statusCode": 404,
}
`;

exports[`apiRequestRoutine should return error response if internal error is thrown in handler 1`] = `
Object {
  "body": "{\\"status\\":\\"ServiceUnavailableError\\",\\"statusCode\\":503,\\"message\\":\\"Upstream service is unavailable\\"}",
  "statusCode": 503,
}
`;

exports[`apiRequestRoutine should return success response 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;
